packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -318,28 +326,64 @@ export class Stack extends Construct implements ITaggable {
   /**
    * Add a dependency between this stack and another stack
    */
-  public addDependency(stack: Stack, reason?: string) {
-    if (stack === this) { return; }  // Can ignore a dependency on self
+  public addDependency(target: Stack, reason?: string) {
+    const source = this;
 
-    reason = reason || 'dependency added using stack.addDependency()';
-    const dep = stack.stackDependencyReasons(this);
-    if (dep !== undefined) {
-        // tslint:disable-next-line:max-line-length
-        throw new Error(`'${stack.node.path}' depends on '${this.node.path}' (${dep.join(', ')}). Adding this dependency (${reason}) would create a cyclic reference.`);
+    if (source === target) { return; }
+
+    const common = findCommonStack(source, target);
+
+    // if there is no common stack (i.e. we got to the root), then define an assembly-
+    // level dependency between the two top-level stacks
+    if (!common) {
+      const topLevelSource = findTopLevelStack(source);
+      const topLevelTarget = findTopLevelStack(target);
+
+      topLevelSource.addAssemblyDependency(topLevelTarget, reason);
+
+      return;
     }
-    this._stackDependencies.add({ stack, reason });
 
-    if (process.env.CDK_DEBUG_DEPS) {
-      // tslint:disable-next-line:no-console
-      console.error(`[CDK_DEBUG_DEPS] stack "${this.node.path}" depends on "${stack.node.path}" because: ${reason}`);
+    // if both source & target are top-level stacks and we have a common stack
+    // by definition it means that source === target == common.
+    if (!target.nested && !source.nested) {
+
+      if (target !== source || source !== common) {
+        throw new Error(`Unexpected: if two top-level stacks have a common stack, they all must be the same stack`);
+      }
+
+      // source & target are the same, ignore
+      return;
+    }
+
+    // nested stacks cannot depend on any of its parent stacks
+    if (target.isParentOfNestedStack(source)) {
+      throw new Error(`Nested stack '${source.node.path}' cannot depend on a parent stack '${target.node.path}'`);
+    }
+
+    // if the target is a nested stack of the source (directly or indirectly),
+    // then we can ignore because by definition it will be deployed before it.
+    if (source.isParentOfNestedStack(target)) {
+      return;
+    }
+
+    // at this point, we know that both source and target are nested stacks
+    if (!source.nestedStackResource || !target.nestedStackResource) {
+      throw new Error(`Unexpected: both '${source.node.path}' and '${target.node.path}' are expected to be nested stacks`);
     }
+
+    // both source and target are nested, so we are going to transfer the dependency to the
+    // respective AWS::CloudFormation::Stack resources that share the same common stack.
+    const sourceResource = findCommonCfnResource(source.nestedStackResource, common);
```

Comment:
`nestedStackResource` is only used for the purposes of `findCommonCfnResource()`. So seems like if we make that a virtual method on `Stack`, we don't need the tight coupling between `core` and `aws-cloudformation` and we can leave `NestedStack` where it is?

```ts
Stack.of(source).findRepresentingResource(source, common);

class Stack {
  public findRepresentingResource(resource: CfnResource, inside: Stack) {
    if (inside !== this) {
      throw new Error('Ruh-roh!');
    }
    return resource;
  }
}

class NestedStack {
  public findRepresentingResource(resource: CfnResource, inside: Stack) {
    if (inside === this) {
      return resource;
    }
    return Stack.of(this).findRepresentingResource(this.nestedStackResource, inside);
  }
}
```

Reply:
Done